### PR TITLE
8320192: SHAKE256 does not work correctly if n >= 137

### PR DESCRIPTION
**Pull request by wangweij**
When blocksize is less than digest output length, call `keccak` more.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320192](https://bugs.openjdk.org/browse/JDK-8320192): SHAKE256 does not work correctly if n &gt;= 137 (**Bug** - P3)

### Reviewers
 * [Mark Powers](https://openjdk.org/census#mpowers) (@mcpowers - Committer) ⚠️ Review applies to [26a6f118](https://git.openjdk.org/jdk/pull/16698/files/26a6f118bd53246ec4fd8eaed50b352344ed0f0e)
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)

### Contributors
 * Ferenc Rakoczi `<ferenc.r.rakoczi@oracle.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16698/head:pull/16698` \
`$ git checkout pull/16698`

Update a local copy of the PR: \
`$ git checkout pull/16698` \
`$ git pull https://git.openjdk.org/jdk.git pull/16698/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16698`

View PR using the GUI difftool: \
`$ git pr show -t 16698`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16698.diff">https://git.openjdk.org/jdk/pull/16698.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16698#issuecomment-1815293505)

### CODE REVIEW

File: src/java.base/share/classes/sun/security/provider/SHAKE256.java [hunk 1]
```diff
@@ -28,21 +28,41 @@
  * The SHAKE256 extendable output function.
  */
 public final class SHAKE256 extends SHA3 {
+    // d is the required number of output bytes
+    // if this constructor is used with d > 0, the squeezing methods will throw a ProviderException
```
Comment (mcpowers) [inline src/java.base/share/classes/sun/security/provider/SHAKE256.java:32]: These should be complete sentences capital letters and periods.

File: src/java.base/share/classes/sun/security/provider/SHAKE128.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+public final class SHAKE128 extends SHA3 {
+    // d is the required number of output bytes
+    // if this constructor is used with d > 0, the squeezing methods will throw a ProviderException
+    public SHAKE128(int d) {
+        super("SHAKE128", d, (byte) 0x1F, 32);
+    }
+
+    // if this constructor is used to get an instance of the class, then, after the last update,
+    // one can get the generated bytes using the squeezing methods
+    // calling digest method will throw a ProviderException
```
Comment (mcpowers) [inline src/java.base/share/classes/sun/security/provider/SHAKE128.java:39]: These should be complete sentences capital letters and periods.

File: src/java.base/share/classes/sun/security/provider/DigestBase.java [hunk 3]
```diff
@@ -137,7 +137,11 @@ protected final void engineUpdate(byte[] b, int ofs, int len) {
         }
     }
 
-    // compress complete blocks
+    protected final void engineUpdate(byte[] buf) {
+        engineUpdate(buf, 0, buf.length);
+    }
+
+        // compress complete blocks
```
Comment (mcpowers) [inline src/java.base/share/classes/sun/security/provider/DigestBase.java:144]: Indentation problem.

**Review by mcpowers — APPROVED** [review]
Looks good. I only found trivial things. A comment describing `squeezeOffset` would help.

**Comment by mcpowers** [comment]
Nice!

**Comment by valeriepeng** [comment]
Changes look fine. The synopsis should be updated as this is about SHAKE128? In addition, now that the digest output can be variable-length, it'd be good to update the comment for the `private void keccak() `method since the capacity c is not twice of digest length as in the case of SHA-3.

**Review by valeriepeng — APPROVED** [review]
Looks good to me. Thanks~

**Pull request closed without merging**